src/hotspot/share/runtime/atomic.hpp

```diff
@@ -1232,4 +1251,492 @@ inline T Atomic::XchgUsingCmpxchg<byte_size>::operator()(T volatile* dest,
+                   atomic_memory_order order = memory_order_conservative) {
+    return Atomic::fetch_then_add(value_ptr(), add_value, order);
+  }
+
+  template<typename I>
+  T sub_then_fetch(I sub_value,
+                   atomic_memory_order order = memory_order_conservative) {
+    return Atomic::sub(value_ptr(), sub_value, order);
+  }
+
+  template<typename I>
+  T fetch_then_sub(I sub_value,
+                   atomic_memory_order order = memory_order_conservative) {
+    // Atomic doesn't currently provide fetch_then_sub.
+    return sub_then_fetch(sub_value, order) + sub_value;
+  }
+
+  void atomic_inc(atomic_memory_order order = memory_order_conservative) {
+    Atomic::inc(value_ptr(), order);
+  }
+
+  void atomic_dec(atomic_memory_order order = memory_order_conservative) {
+    Atomic::dec(value_ptr(), order);
+  }
+};
+
+template<typename T>
+class Atomic::IntegerValue : private ValueSupportsArithmetic<T> {
+  using Base = ValueSupportsArithmetic<T>;
+  using Base::value_ptr;
+
+public:
+  explicit IntegerValue(T value = 0) : Base(value) {}
+
+  NONCOPYABLE(IntegerValue);
+
+  using ValueType = T;
+
+  using Base::value_offset_in_bytes;
+  using Base::value_size_in_bytes;
+  using Base::load_relaxed;
+  using Base::load_acquire;
+  using Base::relaxed_store;
+  using Base::release_store;
+  using Base::release_store_fence;
+  using Base::cmpxchg;
+  using Base::xchg;
+  using Base::add_then_fetch;
+  using Base::fetch_then_add;
+  using Base::sub_then_fetch;
+  using Base::fetch_then_sub;
+  using Base::atomic_inc;
+  using Base::atomic_dec;
+
+  T fetch_then_and(T bits, atomic_memory_order order = memory_order_conservative) {
+    return Atomic::fetch_then_and(value_ptr(), bits, order);
+  }
+
+  T fetch_then_or(T bits, atomic_memory_order order = memory_order_conservative) {
+    return Atomic::fetch_then_or(value_ptr(), bits, order);
+  }
+
+  T fetch_then_xor(T bits, atomic_memory_order order = memory_order_conservative) {
+    return Atomic::fetch_then_xor(value_ptr(), bits, order);
+  }
+
+  T and_then_fetch(T bits, atomic_memory_order order = memory_order_conservative) {
+    return Atomic::and_then_fetch(value_ptr(), bits, order);
+  }
+
+  T or_then_fetch(T bits, atomic_memory_order order = memory_order_conservative) {
+    return Atomic::or_then_fetch(value_ptr(), bits, order);
+  }
+
+  T xor_then_fetch(T bits, atomic_memory_order order = memory_order_conservative) {
+    return Atomic::xor_then_fetch(value_ptr(), bits, order);
+  }
+};
+
+template<typename T>
+class Atomic::ByteValue : private ValueAccess<T> {
+  using Base = ValueAccess<T>;
+
+public:
+  explicit ByteValue(T value = 0) : Base(value) {}
+
+  NONCOPYABLE(ByteValue);
+
+  using ValueType = T;
+
+  using Base::value_offset_in_bytes;
+  using Base::value_size_in_bytes;
+  using Base::load_relaxed;
+  using Base::load_acquire;
+  using Base::relaxed_store;
+  using Base::release_store;
+  using Base::release_store_fence;
+  using Base::cmpxchg;
```

Comment:
I waffled back and forth about these long lists of `using` declarations. Instead maybe macroize?